From 04cebcbd17a9b363616d54ef7124e199b022c48c Mon Sep 17 00:00:00 2001
From: Saravanakumar Duraisamy <quic_saradura@quicinc.com>
Date: Thu, 9 Nov 2023 15:59:17 +0530
Subject: [PATCH] wifi: ath12k: Ignore duplicate status cb received from MHI

FW sending two different interrupt (ex: state change and Execution
Environment(EE) change) for RDDM and MHI trigger status callback two times.
In ATH layer second callbacks interpreted as FW crash during the recovery hence
driver triggering BUG_ON.

Driver ignoring the duplicate event received from MHI based on the
current and privious EE value for RDDM callback because both RDDM callback has
same EE.

Signed-off-by: Saravanakumar Duraisamy <quic_saradura@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/mhi.c | 8 ++++++++
 1 file changed, 8 insertions(+)

--- a/drivers/net/wireless/ath/ath12k/mhi.c
+++ b/drivers/net/wireless/ath/ath12k/mhi.c
@@ -366,6 +366,15 @@ static void ath12k_mhi_op_status_cb(stru
 		ath12k_warn(ab, "firmware crashed: MHI_CB_SYS_ERROR\n");
 		break;
 	case MHI_CB_EE_RDDM:
+		ath12k_warn(ab, "Received MHI callback: MHI_CB_EE_RDDM\n");
+		/* check duplicate RDDM received from MHI */
+		if (mhi_get_exec_env(ab_pci->mhi_ctrl) == mhi_cntrl->ee) {
+			ath12k_dbg(ab, ATH12K_DBG_BOOT, "Ignore duplicate status cb %s(%d) received from MHI\n",
+				    ath12k_mhi_op_callback_to_str(cb),
+				    cb);
+			return;
+		}
+
 		/* In-case of rddm for mhi soc reset */
 		if(test_bit(ATH12K_MHI_SOC_RESET, &ab_pci->mhi_state)) {
 			ath12k_dbg(ab, ATH12K_DBG_BOOT, "Triggering RDDM from mhi soc reset\n");
@@ -376,10 +385,10 @@ static void ath12k_mhi_op_status_cb(stru
 
 		/* In-case of rddm for fatal error and ab is registered. */
 		if (!(test_bit(ATH12K_FLAG_UNREGISTERING, &ab->dev_flags))) {
+			ath12k_info(ab, "Schedule SSR Recovery reset work queue\n");
 			set_bit(ATH12K_FLAG_CRASH_FLUSH, &ab->dev_flags);
 			set_bit(ATH12K_FLAG_RECOVERY, &ab->dev_flags);
 			queue_work(ab->workqueue_aux, &ab->reset_work);
-			ath12k_dbg(ab, ATH12K_DBG_BOOT, "Schedule SSR Recovery reset work queue\n");
 			ath12k_hal_dump_srng_stats(ab);
 		} else {
 			BUG_ON(1);
